src/lib/work_partitioner/work_partitioner.ml

```diff
@@ -242,3 +242,118 @@ let request_partitioned_work ~(sok_message : Mina_base.Sok_message.t)
         (request_from_selector_and_consume_by_partitioner ~partitioner
            ~work_from_selector ~sok_message )
     ]
+
+(* TODO: figure out the correct cache DB *)
+let proof_cache_db = Proof_cache_tag.create_identity_db ()
+
+type submit_result =
+  | SchemeUnmatched
+  | Removed
+  | Processed of Work.Result.Combined.t option
+      (** If the `option` in Processed is present, it indicates we need to submit to the underlying selector *)
+
+let submit_single ~partitioner
+    ~(submitted_result : (unit, Ledger_proof.t) Work.Result.Single.Poly.t)
+    ~job_id =
+  let Work.Id.Single.{ which_one; pairing_id } = job_id in
+  let result = ref SchemeUnmatched in
+  Hashtbl.change partitioner.pairing_pool pairing_id ~f:(function
+    | Some pending_combined_result -> (
+        let submitted_result =
+          Snark_work_lib.Result.Single.Poly.map ~f_spec:Fn.id
+            ~f_proof:(Ledger_proof.Cached.write_proof_to_disk ~proof_cache_db)
+            submitted_result
+        in
+        match
+          Combining_result.merge_single_result pending_combined_result
+            ~submitted_result ~submitted_half:which_one
+        with
+        | Pending pending ->
+            result := Processed None ;
+            Some pending
+        | Done combined ->
+            result := Processed (Some combined) ;
+            None
+        | HalfAlreadyInPool | StructureMismatch _ ->
+            result := SchemeUnmatched ;
+            Some pending_combined_result )
+    | None ->
+        (* We should always at least having a Spec_only in the pool, this branch
+           hence would be SchemeUnmatched indeed *)
+        None ) ;
+  !result
+
+let submit_into_pending_zkapp_command ~partitioner
+    ~(job_id : Work.Id.Sub_zkapp.t)
+    ~data:
+      ({ proof; data = elapsed } :
+        (Core.Time.Span.t, Ledger_proof.t) Proof_carrying_data.t ) =
+  let returns = ref SchemeUnmatched in
+  let process (pending : Pending_zkapp_command.t) =
+    Pending_zkapp_command.submit_proof ~proof ~elapsed pending ;
+
+    match Pending_zkapp_command.try_finalize pending with
+    | None ->
+        returns := Processed None
+    | Some ({ job_id; _ }, proof, elapsed) ->
+        let submitted_result : (unit, Ledger_proof.t) Work.Result.Single.Poly.t
+            =
+          Work.Result.Single.Poly.{ spec = (); proof; elapsed }
+        in
+        returns := submit_single ~partitioner ~submitted_result ~job_id
+  in
```

Comment:
Should we remove the pending zkapp command from the relevant pool here?

Reply:
This is removed because the outer invocation of `change_inplace`